src/depends/NAT/nat.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+	This file is part of cpp-ethereum.
+	cpp-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/// @file
+/// UPnP port forwarding support.
+
+#include "nat.h"
+// #include "libUtils/Logger.h"
+#include <arpa/inet.h> //INET6_ADDRSTRLEN
+#include <cstring>
+#include <iostream>
+#include <miniupnpc/miniupnpc.h>
+#include <miniupnpc/upnpcommands.h>
+#include <miniupnpc/upnperrors.h>
+
+using namespace std;
+
+namespace
+{
+    const int DISCOVERY_TIME_OUT = 2000;
+    const int TTL  = 2;   
+}
+
+NAT::NAT()
+{
+    m_urls = make_shared<UPNPUrls>();
+    m_data = make_shared<IGDdatas>();
+
+    memset(m_urls.get(), 0, sizeof(struct UPNPUrls));
+    memset(m_data.get(), 0, sizeof(struct IGDdatas));
+
+    m_initialized = false;
+
+    shared_ptr<struct UPNPDev> devlist;
+    int error = 0;
+    
+#if MINIUPNPC_API_VERSION >= 14
+        devlist.reset(upnpDiscover(DISCOVERY_TIME_OUT, NULL/*multicast interface*/, NULL/*minissdpd socket path*/, 0/*sameport*/, 0/*ipv6*/, TTL, &error));
+#else
+        devlist.reset(upnpDiscover(DISCOVERY_TIME_OUT, NULL/*multicast interface*/, NULL/*minissdpd socket path*/, 0/*sameport*/, 0/*ipv6*/, &error));
+#endif
+
+    if (devlist.get() == NULL || error != 0)
+    {
+        return;
+    }
+
+    char lan_address[INET6_ADDRSTRLEN];
+
+    int status = UPNP_GetValidIGD(devlist.get(), m_urls.get(), m_data.get(),
+                                  lan_address, sizeof(lan_address));
+
+    if (status != 1)
+    {
+        return;
+    }
+
+    m_lanAddress = lan_address;
+    m_initialized = true;
+}
+
+NAT::~NAT()
+{
+    auto r = m_reg;
```

Comment:
So this is tricky. Clever that you do a copy, because you iterate `m_reg`, while removing from it using `removeRedirect`. As a naive maintainer, I would be tempted to remove it, optimize it though. So, I think I would prefer it to have a more intention seeking name e.g. `r` -> `regCopySafeToIterateDuringDelete` or something like that.